src/resolve/resolved-dns-answer.c

```diff
@@ -685,6 +686,19 @@ int dns_answer_reserve(DnsAnswer **a, size_t n_free) {
                         return -ENOMEM;
 
                 n->n_allocated = ns;
+
+                set_clear(n->set_items);
+                r = set_reserve(n->set_items, ns);
+                if (r < 0)
+                        return r;
+
+                for (size_t i = 0; i < n->n_rrs; i++) {
+                        r = set_put(n->set_items, &n->items[i]);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                return -EEXIST;
+                }
```

Comment:
so, previously the function had the nice property that it either succeeded or it failed entirely leaving the object the way it was. With this rework, if the realloc() worked, but the set_reserve fails, then we'll have an invalid object.

If you move the set_reserve() before the realloc() then we should be good, as in that case the set_put() calls are guaranteed to succeed.

hence, please do set_reserve() first, then the realloc(), then the set_clear(), and then the set_put() loop. The latter should be enclosed in `assert_se()`, since the earlier set_reserve() should have given us the guarantee that the set_put() will work (and given that the elements previously fit in the same set, hence they should still be good to fit, without conflicts, without EEXIST).

Reply:
and of course add a brief comment next to the assert_se() explaining why we get the guarantee that the set_put() will succeed, since assert_se()'ing around something that can theoretically fail will look like a bug at first sight.